src/caf/core/data_structures.py

```diff
@@ -706,5 +1448,292 @@ def old_to_new_dvec(import_data: dict):
+                zone = balancing_zones._segment_zoning[seg]
+                self_slice = self.filter_segment_value(seg, vals)
+                self_remaining = self.drop_by_segment_values(seg, vals)
+                other_slice = other.filter_segment_value(seg, vals)
+                other_remaining = other.drop_by_segment_values(seg, vals)
+                slice_factors = self._balance_zones_internal(
+                    self_slice, self.zoning_system, other_slice, other.zoning_system, zone
+                )
+                remaining_factors = self._balance_zones_internal(
+                    self_remaining,
+                    self.zoning_system,
+                    other_remaining,
+                    other.zoning_system,
+                    balancing_zones._default_zoning,
+                )
+                balanced = pd.concat(
+                    [self_slice * slice_factors, self_remaining * remaining_factors]
+                )
+            else:
+                balanced = self.data.copy()
+                for zon, segs in balancing_zones.zoning_groups():
+                    grouped_self = self.data.groupby(level=segs).sum()
+                    grouped_other = other.data.groupby(level=segs).sum()
+                    factors = self._balance_zones_internal(
+                        grouped_self,
+                        self.zoning_system,
+                        grouped_other,
+                        other.zoning_system,
+                        zon,
+                    )
+                    balanced *= factors
+                remaining_segs = set(self.segmentation.names) - set(
+                    balancing_zones._segment_zoning.keys()
+                )
+                grouped_self = self.data.groupby(level=list(remaining_segs)).sum()
+                grouped_other = other.data.groupby(level=list(remaining_segs)).sum()
+                factors = self._balance_zones_internal(
+                    grouped_self,
+                    self.zoning_system,
+                    grouped_other,
+                    other.zoning_system,
+                    balancing_zones._default_zoning,
+                )
+                balanced *= factors
+                # Factors have been applied multiple times at different segmentation levels, so need to balance once more over whole rows to make totals match
+                factor = other.data.sum(axis=1) / balanced.sum(axis=1)
+                balanced *= factor
+        else:
+            raise ValueError(
+                "balancing_zones must be either BalancingZones, ZoningSystem, or None"
+                f"type provided: {type(balancing_zones)}"
+            )
+        return DVector(
+            import_data=balanced,
+            segmentation=self.segmentation,
+            zoning_system=self.zoning_system,
+            time_format=self.time_format,
+        )
+
+
+@dataclass
+class IpfTarget:
+    """Dataclass to store targets to pass to IPF method of DVector."""
```

Comment:
Could the docstring explaing what the attributes are as I don't think it's completely obvious from the names. Also, could a name attribute be added to help with reporting it would potentially just be used to distinguish productions, attractions etc...